### PR TITLE
PP-611 normalize create charge api errors

### PR DESCRIPTION
**Pull request by rauligs**
- Create charge resource is now normalised for the different errors (Json structure and validations) with the codes defined in the existing document [Public API](https://docs.google.com/document/d/1I9HZRdG5DjiUG96sjiMojVGZ5EIl7Sl9ORM2d93yJao/edit)

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/app/PublicApi.java [hunk 1]
```diff
@@ -35,17 +42,27 @@ public void initialize(Bootstrap<PublicApiConfig> bootstrap) {
...
+
+    private void configureObjectMapper(PublicApiConfig config, ObjectMapper objectMapper) {
+
+        URLValidator urlValidator = urlValidatorValueOf(config.getRestClientConfig().isDisabledSecureConnection());
+        CreatePaymentRequestDeserializer paymentRequestDeserializer = new CreatePaymentRequestDeserializer(new PaymentRequestValidator(urlValidator));
+
+        SimpleModule customDeserializationModule = new SimpleModule("customDeserializationModule");
+        customDeserializationModule.addDeserializer(CreatePaymentRequest.class, paymentRequestDeserializer);
+
+        objectMapper.configure(DeserializationFeature.ACCEPT_FLOAT_AS_INT, false);
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/api/app/PublicApi.java]: 👍

File: src/main/java/uk/gov/pay/api/exception/mapper/CreateChargeConnectorErrorResponseExceptionMapper.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CreateChargeConnectorErrorResponseException.class);
+
+    @Override
+    public Response toResponse(CreateChargeConnectorErrorResponseException exception) {
+
+        int errorStatus = exception.getErrorStatus();
+
+        if (errorStatus == NOT_FOUND.getStatusCode()) {
+            LOGGER.error("Authorization succeeded but Connector response was Gateway account not found: {}", exception.getErrorBody());
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/api/exception/mapper/CreateChargeConnectorErrorResponseExceptionMapper.java]: Have these messages vetted by designers/user researchers? May be these are good candidate for place in a resource(properties) file.

Comment (rauligs) [inline src/main/java/uk/gov/pay/api/exception/mapper/CreateChargeConnectorErrorResponseExceptionMapper.java]: There is a story in the backlog for logging.
And for error messages back to the user I'll mention it in stand up to see if they are interested on them

File: src/main/java/uk/gov/pay/api/json/CustomParser.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package uk.gov.pay.api.json;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import uk.gov.pay.api.exception.BadRequestException;
+import uk.gov.pay.api.model.PaymentError;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+import static uk.gov.pay.api.model.PaymentError.*;
+
+class CustomParser {
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/api/json/CustomParser.java]: Custom *what* parser? / Perhaps change the name to something more specific

File: src/main/java/uk/gov/pay/api/model/PaymentError.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    public static PaymentError invalidAttributeValue(Code code, String fieldName, String explanation) {
+        return new PaymentError(code, "Invalid attribute value: " + fieldName + ". " + explanation);
+    }
+
+    @ApiModelProperty(example = "P0198")
+    public String getCode() {
+        return code.name();
+    }
+
+    @ApiModelProperty(example = "Downstream system error")
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/api/model/PaymentError.java]: May be not a the best example error for integrators.

Comment (rauligs) [inline src/main/java/uk/gov/pay/api/model/PaymentError.java]: True, changed in the next pull request

File: src/main/java/uk/gov/pay/api/json/CustomParser.java (hunk 3 above)
Comment (rauligs) [inline src/main/java/uk/gov/pay/api/json/CustomParser.java]: Will be changed to Json parser

File: src/test/java/uk/gov/pay/api/it/PaymentsResourceAmountValidationITest.java [hunk 5]
```diff
@@ -95,14 +120,21 @@ public void createPayment_responseWith400_whenAmountFieldIsNotAValidJsonField()
                 "  \"return_url\" : \"http://somewhere.gov.uk/rainbow/1\"" +
                 "}";
 
-        postPaymentResponse(BEARER_TOKEN, payload)
+        System.out.println("payload = " + payload);
```
Comment (kakumara) [inline src/test/java/uk/gov/pay/api/it/PaymentsResourceAmountValidationITest.java]: sysout ??

**Comment by kakumara** [comment]
Otherwise looks 👍

File: src/test/java/uk/gov/pay/api/it/PaymentsResourceAmountValidationITest.java (hunk 5 above)
Comment (rauligs) [inline src/test/java/uk/gov/pay/api/it/PaymentsResourceAmountValidationITest.java]: Will remove it! ups

**Pull request merged**